### PR TITLE
client/v3/watch.go: use `fmt` go pkg for metadata map printing

### LINKED ISSUE #18303: Code that relies on gRPC metadata formatting should be fixed
### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via [etcd maintainers](mailto:etcd-maintainers@googlegroups.com).
- [X] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

This code:

https://github.com/etcd-io/etcd/blob/21e5876f7f0539509c277b4c2a3ba1b1599b1721/client/v3/watch.go#L1038-L1043

expects consistent ordering from gRPC as well as expecting all metadata to be included.

However, metadata can contain PII, so we would like to change gRPC's behavior to not print any potential PII.  https://github.com/grpc/grpc-go/issues/7395

### What did you expect to happen?

Code correctness to _not_ rely upon gRPC's `metadata.MD.String` method.

### How can we reproduce it (as minimally and precisely as possible)?

Look at the code.

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

master

### Etcd configuration (command line flags or environment variables)

_No response_

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by mohamedawnallah**
## Description

Addresses this issue #18303

### CODE REVIEW

File: client/v3/watch.go [hunk 1]
```diff
@@ -1037,7 +1037,7 @@ func (pr *progressRequest) toPB() *pb.WatchRequest {
 
 func streamKeyFromCtx(ctx context.Context) string {
 	if md, ok := metadata.FromOutgoingContext(ctx); ok {
-		return fmt.Sprintf("%+v", md)
+		return fmt.Sprint(map[string][]string(md))
```
Comment (ahrtr) [inline client/v3/watch.go:1040]: let's try to keep the format consistent
```suggestion
		return fmt.Sprintf("%+v", map[string][]string(md))
```

**Comment by ahrtr** [comment]
Please also add an unit test

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -52,7 +52,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: get-image
-        uses: actions/download-artifact@65a9edc5881444af0b9093a5e628f2fe47ea3b2e # v4.1.7
```
Comment (henrybear327) [inline .github/workflows/release.yaml:55]: Is this change intended?

Comment (mohamedawnallah) [reply .github/workflows/release.yaml]: No. just rebased on main branch 👍

File: client/v3/watch_test.go [hunk 3]
```diff
@@ -15,9 +15,11 @@
 package clientv3
 
 import (
+	"context"
 	"testing"
 
 	"go.etcd.io/etcd/api/v3/mvccpb"
+	"google.golang.org/grpc/metadata"
```
Comment (ahrtr) [inline client/v3/watch_test.go:22]: This should can fix the workflow failure https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/etcd-io_etcd/18308/pull-etcd-verify/1810975417960501248
```suggestion
	"google.golang.org/grpc/metadata"
	
	"go.etcd.io/etcd/api/v3/mvccpb"
```

File: client/v3/watch_test.go [hunk 4]
```diff
@@ -53,3 +55,53 @@ func TestEvent(t *testing.T) {
...
+			md:       metadata.MD{},
+			expected: "map[]",
+		},
+		{
+			md: metadata.MD{
+				"key1": []string{"value1", "value1a"},
+			},
+			expected: "map[key1:[value1 value1a]]",
+		},
+	}
```
Comment (ahrtr) [inline client/v3/watch_test.go:88]:
```suggestion
	tests := []struct {
		name  string
		ctx      context.Context
		expected string
	}{
		{
			name: "multiple keys",
			ctx: metadata.NewOutgoingContext(context.Background(), metadata.MD{
				"key1": []string{"value1"},
				"key2": []string{"value2a", "value2b"},
			}),
			expected: "map[key1:[value1] key2:[value2a value2b]]",
		},
		{
			name: "no keys",
			ctx:       metadata.NewOutgoingContext(context.Background(), metadata.MD{}), 
			expected: "map[]",
		},
		{
			name: "only one key",
			ctx: metadata.NewOutgoingContext(context.Background(), metadata.MD{
				"key1": []string{"value1", "value1a"},
			}),
			expected: "map[key1:[value1 value1a]]",
		},
		{
			name: "no metadata",
			ctx:       context.Background(),
			expected: "",
		},		
	}
```

File: client/v3/watch_test.go [hunk 5]
```diff
@@ -53,3 +55,53 @@ func TestEvent(t *testing.T) {
...
+		}
+	}
+
+	// Test with no metadata in context to ensure it returns an empty string.
+	ctx := context.Background()
+	expected := ""
+	actual := streamKeyFromCtx(ctx)
+	if actual != expected {
+		t.Errorf("streamKeyFromCtx() = %v, want %v", actual, expected)
+	}
```
Comment (ahrtr) [inline client/v3/watch_test.go:106]:
```suggestion
```

File: client/v3/watch_test.go [hunk 6]
```diff
@@ -53,3 +55,53 @@ func TestEvent(t *testing.T) {
...
+
+	// Iterate through each test case and verify the function's output matches
+	// the expected result.
+	for i, tt := range tests {
+		ctx := metadata.NewOutgoingContext(context.Background(), tt.md)
+		actual := streamKeyFromCtx(ctx)
+		if actual != tt.expected {
+			t.Errorf("#%d: streamKeyFromCtx() = %v, want %v", i, actual, tt.expected)
+		}
+	}
```
Comment (ahrtr) [inline client/v3/watch_test.go:98]: Pls use subtest, 

```
t.Run(tt.name, func(t *testing.T) {
    ......
})
```

For example, 
https://github.com/etcd-io/etcd/blob/4d9dcebd61d63060edd5428c834d01921a03071f/server/etcdserver/api/v3discovery/discovery_test.go#L79-L101

**Comment by mohamedawnallah** [comment]
Thanks a lot for the feedback, @ahtr! It is much better now! 🙏 I have addressed all of the feedback.

File: client/v3/watch_test.go [hunk 7]
```diff
@@ -53,3 +56,59 @@ func TestEvent(t *testing.T) {
...
+		},
+		{
+			name:     "no metadata",
+			ctx:      context.Background(),
+			expected: "",
+		},
+	}
+
+	// Iterate through each test case and verify the function's output matches
+	// the expected result.
```
Comment (ahrtr) [inline client/v3/watch_test.go:101]: The test is self-explanation, it's also simple enough, no need to add comment.
```suggestion
```

File: client/v3/watch_test.go [hunk 8]
```diff
@@ -53,3 +56,59 @@ func TestEvent(t *testing.T) {
...
+	for _, tt := range tests {
+		// Capture range variable to ensure each closure gets its own copy of tt.
+		tt := tt
```
Comment (ahrtr) [inline client/v3/watch_test.go:104]: FYI. golang 1.22 already resolved this issue (variables declared by a “for” loop overwritten by next iteration), it isn't necessary anymore starting from 1.22. But feel free to keep it as it's.

File: client/v3/watch_test.go [hunk 9]
```diff
@@ -53,3 +56,59 @@ func TestEvent(t *testing.T) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			// Run tests in parallel.
+			t.Parallel()
```
Comment (ahrtr) [inline client/v3/watch_test.go:107]: Let's keep it as simple as possible.
```suggestion
```

**Comment by mohamedawnallah** [comment]
I've addressed the recent feedback 🙏. This might be off-topic, but I'm eager to contribute more to `etcd`. Could you suggest any specific issues or labels to work on? I'm just starting out and haven't found any labeled `good-first-issue` other than this one.

**Comment by ahrtr** [comment]
> Could you suggest any specific issues or labels to work on?

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md

**Review by ahrtr — APPROVED** [review]
LGTM

nice work! Thank you for your first contribution!

**Review by ivanvc — COMMENTED** [review]
Looking good. Thanks, @mohamedawnallah. I left a potential suggestion.

I also suggest removing the "Close" keyword from the pull request description, as Benjamin labeled the issue with a backport to 3.4 and 3.5, and it would be better not to close the parent issue automatically.

File: client/v3/watch_test.go [hunk 10]
```diff
@@ -53,3 +56,53 @@ func TestEvent(t *testing.T) {
...
+			ctx: metadata.NewOutgoingContext(context.Background(), metadata.MD{
+				"key1": []string{"value1", "value1a"},
+			}),
+			expected: "map[key1:[value1 value1a]]",
+		},
+		{
+			name:     "no metadata",
+			ctx:      context.Background(),
+			expected: "",
+		},
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply client/v3/watch_test.go:97]: Not a big deal, recommend retaining these two test cases to ensure the completeness of the test cases.

**Review by ivanvc — APPROVED** [review]
LGTM. We can add completeness in a follow-up/later. Thanks.

**Comment by ahrtr** [comment]
@mohamedawnallah could you backport the PR to 3.5 and 3.4?

**Comment by mohamedawnallah** [comment]
Yeah I would love that it is my first time going through this backport process! I've just submitted two PRs for this:
- backport to release `3.4` PR: #18311
- backport to release `3.5` PR: #18312

**Pull request merged**